hotstuff/src/committer.rs

```diff
@@ -0,0 +1,269 @@
+use crate::consensus::{Round, CHANNEL_CAPACITY};
+use crate::error::{ConsensusError, ConsensusResult};
+use config::Committee;
+use crypto::Hash as _;
+use crypto::{Digest, PublicKey};
+use futures::future::try_join_all;
+use futures::stream::FuturesOrdered;
+use futures::stream::StreamExt as _;
+use log::{debug, error, info, log_enabled};
+use primary::Certificate;
+use std::cmp::max;
+use std::collections::{HashMap, HashSet};
+use store::Store;
+use tokio::sync::mpsc::{channel, Receiver, Sender};
+
+/// The representation of the DAG in memory.
+type Dag = HashMap<Round, HashMap<PublicKey, (Digest, Certificate)>>;
+
+/// The state that needs to be persisted for crash-recovery.
+struct State {
+    /// The last committed round.
+    last_committed_round: Round,
+    // Keeps the last committed round for each authority. This map is used to clean up the dag and
+    // ensure we don't commit twice the same certificate.
+    last_committed: HashMap<PublicKey, Round>,
+    /// Keeps the latest committed certificate (and its parents) for every authority. Anything older
+    /// must be regularly cleaned up through the function `update`.
+    dag: Dag,
+}
+
+impl State {
```

Comment:
The interface Primary<-> Consensus seems mixed. We use the store in different ways (certificate_waiter, state_update). Might be worth having a clear interface on 1) How to parse the DAG from the primary and 2)How to return GC_Round back. 

At the old code the interface was a bit deeper (only the primary new the DAG and did the final ordering),  not sure which is better.